### ISSUE TITLE
Could spring-boot-starters enable requireUpperBoundDeps instead of dependencyConvergence?

### ISSUE THREAD

**Issue by cpovirk** [spring-projects/spring-boot#14275]
`dependencyConvergence`, [enabled](https://github.com/spring-projects/spring-boot/blob/f3dbd946dec9e10e2737ef7113fefb38c862f609/spring-boot-project/spring-boot-starters/pom.xml#L93) by `spring-boot-starters` [since 2014](https://github.com/spring-projects/spring-boot/commit/c719ab7aeeed8d3b8fe2c6d02fb3170e85346351#diff-1360dc44d94c0413fb48c5f98b7b9025), is a very strict property. I would argue that it is:

1. stricter than what users need
2. difficult to achieve, especially for libraries

＃1: What do users want? IIUC, what they generally want is for Maven to resolve the correct version of a library. They don't necessarily mind if other versions are _present_ in the graph, provided that those versions will never be chosen by Maven. But `dependencyConvergence` requires that no other versions even be present. That is, all other versions must be overridden with `dependencyManagement` or removed with `exclusions`.

＃2: Why is it hard for libraries? It's not actually hard for a library to add a `dependencyConvergence` rule and get it to pass: It just needs to add `dependencyManagement` for any conflicting transitive deps. However: This does not help the library's users, since its `dependencyManagement` of transitive deps does not affect them. For a library to present a `dependencyManagement`-clean graph to its users, it must [rely on `exclusions`](https://github.com/grpc/grpc-java/pull/3381/files). This is verbose and error-prone: Whenever the library updates one of its own deps, it should really clear all its exclusions, revisit which versions are requested for each artifact, and then rewrite its exclusions to achieve that.[*]

I propose to enable [`requireUpperBoundDeps`](https://maven.apache.org/enforcer/enforcer-rules/requireUpperBoundDeps.html) instead: Generally speaking, it achieves what users want, and library owners hear complaints only when their libraries are actually resolving a different version than users want.

As best I can tell, the main downside to `requireUpperBoundDeps` for a Spring apps is that it can sometimes be wrong in its own way: The newest version requested in the dependency graph might not actually work with all other libraries. (This can happen when libraries make breaking changes -- which we'd rather they didn't, but obviously they do sometimes.) However, this is fundamentally the same problem as `dependencyConvergence` has: Maven or the user picks a version, and it might or might not be the right one. I think that `requireUpperBoundDeps` is more likely to be correct by default. Plus, I expect it to preserve correctness better over time: Once a user specifies a `foo-1.0` in `dependencyManagement`, `dependencyConvergence` will never issue another error about it, even if the user later upgrades other dependencies to versions that require `foo-1.1`. `requireUpperBoundDeps`, by contrast, will ([usually...](https://issues.apache.org/jira/browse/MENFORCER-316)) issue an error when the new dependency on `foo-1.1` is added -- unless, of course, Maven happens to choose that version by default, in which case the user doesn't have to do anything!

(If a project does need to manually select an older version, it can use `excludes`.)

`requireUpperBoundDeps` fundamentally seems like a good thing to me. It's roughly a way of approximating Gradle's dependency resolution, choosing the newest version rather than Maven's choice of "nearest." It didn't exist when `dependencyConvergence` was first around, so it doesn't seem to have become part of the typical Maven best practices. If you buy my logic, Spring has an opportunity to make it part of the best practices. I, as a library owner, and as a general Maven user, would love to see that happen. Let me know if there's anything I can do to be more persuasive. Thanks for your time.

[*] Even then, users might hypothetically have problems: Suppose that mylib depends directly on foo-1.1 (which depends on bar-1.1) and directly on bar-1.2. It excludes bar from its foo dependency. Now some user of mylib decides to dependency-manage foo to foo-1.3, which fixes a bug. But suppose foo-1.3 has updated its dependency to bar-1.3. Maven will not pull in bar-1.3 because it's referenced only through the excluded foo -> bar chain. I've never seen this happen (not that I've used exclusions much), but I see it as another piece of evidence that exclusions are conceptually the wrong tool.

**Comment by cpovirk** [comment]
Correction / Friday-afternoon parting thought:

> However, this is fundamentally the same problem as `dependencyConvergence` has: Maven or the user picks a version, and it might or might not be the right one.

That's not quite fair: `dependencyConvergence` _forces the user_ to pick a version. Maybe that's good (if you think that's commonly necessary), or maybe it's bad (if you think the user is generally going to start by blindly trying the newest version, anyway).

To summarize behavior for this particular situation:

1. If the graph agrees on the version of a dependency, both `dependencyConvergence` and `requireUpperBoundDeps` pass.
2. If the graph disagrees on the version of a dependency and Maven resolves an older version, both `dependencyConvergence` and `requireUpperBoundDeps` complain.
3. If the graph disagrees on the version of a dependency and Maven resolves the newest version, `dependencyConvergence` complains, but `requireUpperBoundDeps` passes.

So much of the question is about whose behavior is better for ＃3.

**Comment by philwebb** [comment]
The `dependencyConvergence` plugin should only be used by us, it shouldn't be part of a users build. AFAIKT The `<plugins>` section of the POM only gets applied for inherited POMs, not imported ones. If you look at `spring-boot-starter-parent` (the one that most users extend), you'll see it [inherits from spring-boot-dependencies](https://github.com/spring-projects/spring-boot/blob/f3dbd946dec9e10e2737ef7113fefb38c862f609/spring-boot-project/spring-boot-starters/spring-boot-starter-parent/pom.xml#L7) and not `spring-boot-starters`.

Are you finding the `dependencyConvergence` is being somehow applied to your own projects? Can you share an example if so?

**Comment by cpovirk** [comment]
Sorry, shame on me for filing without more research! Let me overcommunicate here and then talk to a couple people before I make any more claims. (Paging **ejona86**, who is subscribed to this issue already.)

I thought I had heard independently from two people that Spring's default configuration did this for user projects. However, the first conversation was a while ago and in real life, so I may have misheard or misremembered, and the second conversation mentioned "users" but didn't specifically say "users other than the people working on the Spring project itself." So I may have read too much into what some people said. Or my sources might have been mistaken (and possibly not as independent as I'd thought :)).

(And sure enough, when I went to file this issue, I went looking in the poms that I expected to have the `<dependencyConvergence/>` line in them, I didn't see it! I ended up just searching the Spring project for that line, and that's how I turned up this probably irrelevant pom.)

It's possible that what's actually going here is this: I talked to 2 people who both maintain projects that are in Spring's dependency tree. (One is on gRPC; the other is on some Google Cloud project.) They'd heard from Spring developers who were annoyed that their projects didn't expose `dependencyConvergence`-compliant surfaces. Maybe those Spring developers should have solved the problems themselves but didn't know that. Maybe the communication between them and the Spring developers wasn't clear, and they misunderstood the scope of the problem. Or maybe something else is going on. I'll see what I can turn up, and I'll let you know what I find.

Thanks for the response, and sorry again for not doing my homework.

**Comment by ejona86** [comment]
It seems this issue can be closed. Thank you!

I mis-understood the documentation I think. When I read "include in your application" I thought it meant "copy it". I see now that's the difference between the "samples" and the "starters." If you copied the starter, it would yield poor results. And the starters appear to be gone anyway starting with the 2.0 release this year.

The initial instance this came up is https://github.com/grpc/grpc-java/issues/1421#issuecomment-185874644 . I had interpreted it as "this is true for all spring boot users" and I did see the enforcer coming up through the parent pom of the starters. I can _now_ interpret it as maybe "Spring thought this was a good idea, so I want to do it too." I had some other discussions with that user where I understood it was a common practice for Spring users, but now I can understand it is self-imposed. That changes things dramatically.

I did a spot-check for a few more of the instances where this came up in grpc-java and google-cloud-java and it was either unclear why they were using `dependencyConvergence` or it was self-imposed.

Sorry @cpovirk for the misunderstanding and @philwebb for consuming your time. But this is great news.
